src/qiboml/operations/differentiation.py

```diff
@@ -152,7 +82,102 @@ def circuit(nqubits = 1):
 
     circuit.set_parameters(original)
 
-    # float() necessary to not return a 0-dim ndarray
-    result = float(generator_eigenval * (forward - backward) * scale_factor)
+    return float(generator_eigenval * (forward - backward))
+
+
+def parameter_shift(
+    hamiltonian: qibo.hamiltonians.Hamiltonian,
+    circuit: qibo.Circuit,
+    exec_backend: qibo.backends.Backend,
+    initial_state: Optional[Union[np.ndarray, qibo.Circuit]] = None,
+    nshots: int = None,
+):
+    """
+    Compute the gradient of the expectation value of ``hamiltonian`` over the
+    state we get executing ``circuit`` over ``initial_state`` with a certain
+    ``nshots`` w.r.t. the parameters of the circuit via parameter-shift
+    rule (https://arxiv.org/abs/1811.11184).
+    The number of shots can be set to be ``None`` in case of exact simulation,
+    otherwise an integer number of shots has to be provided.
+
+    Args:
+        circuit (:class:`qibo.models.circuit.Circuit`): custom quantum circuit.
+        hamiltonian (:class:`qibo.hamiltonians.Hamiltonian`): target observable.
+            if you want to execute on hardware, a symbolic hamiltonian must be
+            provided as follows (example with Pauli Z and ``nqubits=1``):
+            ``SymbolicHamiltonian(np.prod([ Z(i) for i in range(1) ]))``.
+        execution_backend (qibo.backends.Backend): Qibo backend on which the
+            circuits are executed.
+        initial_state (ndarray, optional): initial state on which the circuit
+            acts. Default is ``None``.
+        nshots (int, optional): number of shots if derivative is evaluated on
+            hardware. If ``None``, the simulation mode is executed.
+            Default is ``None``.
+
+    Returns:
+        (float): Gradient of the expectation value of the hamiltonian
+            with respect to circuit's variational parameters.
+
+    """
+    nparams = len(circuit.get_parameters())
+    gradient = []
+
+    for i in range(nparams):
+        gradient.append(
+            _one_parameter_shift(
+                hamiltonian=hamiltonian,
+                circuit=circuit,
+                parameter_index=i,
+                initial_state=initial_state,
+                nshots=nshots,
+                exec_backend=exec_backend,
+            )
+        )
+
+    return gradient
+
+
+def symbolical(
```

Comment:
I had a look at the `tf.autodiff` module, but it seems that the only thing other than `GradientTape` is a [`ForwardAccumulator`](https://www.tensorflow.org/api_docs/python/tf/autodiff/ForwardAccumulator) (so, unexpectedly, TensorFlow can also differentiate in forward mode, though it is not advertising it that much...).

However, it seems that the only way to trace the graph to compute the gradient is evaluating the function with a not-so-dummy value (i.e. an actual value), that is used at the same time as a placeholder and as the value to compute the actual function.

I haven't found the way to decouple the two things.

The only thing we could save is the `circuit.get_parameters()`, that could all replaced with 0s. But we need to know how many of them...

Reply:
> The only thing we could save is the `circuit.get_parameters()`, that could all replaced with 0s. But we need to know how many of them... (and possibly the whole layout, since it's using a list of tuples...)

I hope I'll be able to improve on this in `qibo-core`, https://github.com/qiboteam/qibo-core/issues/22.
But I'm just advertising that issue to trigger your feedback, the benefit of not using the actual parameters in this case is mostly conceptual. It would be nice to avoid doing the work twice, but we also want a TensorFlow independent way to evaluate the function in `expectation`.
So, for the time being, we are paying the price of evaluating twice...

Maybe, in this case, JAX/autograd is truly superior. Though it depends on the evaluation overhead, of course.

Reply:
I just added the jax version of this function. Even if it is ideally better, right now TF is faster.

Reply:
It's quite possible that, even ideally, the speedup is rather limited. Evaluating the function could be cheaper than evaluating the gradient (because the gradient has an extra dimension on the derivation variables).

Still, it's pretty much unexpected that one is slower than the other...

Are you comparing TF-derivation-with-TF-backend, i.e. `(TF, sym, TF)`, with `(JAX, sym, JAX)`?
It could be that the slowdown is just the same of the JAX backend execution, and nothing new.